src/renderers/dom/shared/ReactDOMComponent.js

```diff
@@ -277,7 +283,15 @@ ReactDOMComponent.Mixin = {
   mountComponent: function(rootID, transaction, context) {
     this._rootNodeID = rootID;
 
-    assertValidProps(this, this._currentElement.props);
+    var props = this._currentElement.props;
+
+    var wrapper = wrappers[this._tag];
+    if (wrapper) {
+      wrapper.mountWrapper.call(this, props, context);
```

Comment:
Yea, that's what I meant.

They have the same shape so the function probably does not need to be compiled in a polymorphic way. However, you still have to guard the type coming out + do a lookup on the object of which code path to take. You can't inline it without knowing all the objects that might be here. Which is possible but at the very least you'll end up with a guard.

In my proposal it would be inlined and so you would just immediately know which code path to take.

I'd like to change to use this everywhere. In fact, I would like to change the mountComponent/receiveComponent API between components to do the same. So that it can switch code paths based on `typeof type === 'string'` alone. That's the purpose of the ReactReconciler module - to do the switching.

If the switch cases are very uniform I would recommend just breaking out the switch in a common module.

```
function mountWrapper(inst, props, context) {
  switch (inst._tag) {
    case 'input': ReactDOMInput.mountWrapper(inst, props, context); break;
    ...
  }
}
```

Then just call it:

```js
mountWrapper(inst);
```

This boilerplate actually ends up being smaller than the alternative when you compile it.

Since they're only used in one place, the switch statement and the content of the other modules will just inline right here. As opposed to having an object per module + compressed identifiers to identify each property + the extra `function() { }` etc.

I'm definitely playing human compiler here but the tools aren't doing it for us... yet.

IMO it also makes it easier to follow since every code path is guaranteed. You know that no foreign objects can be introduced. Just follow the stack.